### PR TITLE
Network-action automatons

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the new behavior (if this is a feature change)?**
Simulate network-action automatons by applying them after a contingency or constraint

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
Yes, AUTO network remedial actions and CNECs are now supported

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java [hunk 1]
```diff
@@ -30,46 +32,68 @@
...
+        switch (optimizationState) {
+            case INITIAL:
+                throw new FaraoException("No PerimeterResult for INITIAL optimization state");
+            case AFTER_PRA:
+                return postPreventiveResult;
+            case AFTER_ARA:
+                if (state.getInstant().equals(Instant.CURATIVE)) {
+                    Optional<State> autoState = postContingencyResults.keySet().stream()
+                        .filter(optimizedState -> optimizedState.getInstant().equals(Instant.AUTO) && optimizedState.getContingency().equals(state.getContingency()))
+                        .findFirst();
```
Comment (baptsegu) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java:86]: findAny() is more efficient than findFirst()

File: ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/SearchTreeRaoLoggerTest.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    State statePreventive;
+    State stateCo1Auto;
+    State stateCo1Curative;
+    State stateCo2Curative;
+    String cnec1Id;
+    String cnec2Id;
+    String cnec3Id;
+    String cnec4Id;
+    String cnec5Id;
+    OptimizationResult basecaseOptimResult;
```
Comment (baptsegu) [inline ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/SearchTreeRaoLoggerTest.java]: all private

File: ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/PerimeterOutputTest.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    RangeActionResult prePerimeterRangeActionResult;
+    OptimizationResult optimizationResult;
+    RangeAction ra1;
+    RangeAction ra2;
+    FlowCnec flowCnec1;
+    FlowCnec flowCnec2;
+    NetworkAction na1;
+    NetworkAction na2;
+    PstRangeAction pst1;
+    PstRangeAction pst2;
```
Comment (baptsegu) [inline ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/PerimeterOutputTest.java:48]: all private :p

**Review by baptsegu — APPROVED** [review]
some small comments, and two bigger thoughts (but for other PRs I guess): 

1. It would be more efficient if, depending on the results, we do not store in RaoResult/ write in JSON systematically the ElementaryFlowCnecResult.

For instance, if the cnec is preventive, the only relevant results are INITIAL, AFTER_PRA.
If the cnec is curative, but there is no auto state for the contingency, the only relevant results are INITIAL, AFTER_PRA and AFTER_CRA ...

We can do this to store the data using less memory, while keeping the methods working on the other OptimizationState, for instant, getFlow() on AFTER_CRA/AFTER_ARA returns the AFTER_PRA results for a preventive Cnec.

2. It would be nice to find a clean way to decompose the SearchTreeRaoProvider into several .java files, to makes it easier to read/review :bomb: 


I still need to some to review the validation tests

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java (hunk 1 above)
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java:86]: changed

File: ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/SearchTreeRaoLoggerTest.java (hunk 2 above)
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/SearchTreeRaoLoggerTest.java]: already changed

File: ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/PerimeterOutputTest.java (hunk 3 above)
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/PerimeterOutputTest.java:48]: changed

**Review by baptsegu — APPROVED** [review]

**Pull request merged**